anyfs-tools/doc/man/man8/build_it.8
.TH build_it 8 "19 Jan 2007" "Version 0.84.12d"
.SH "NAME"
build_it \- reads directory entries to obtain information about the
physical location of files on a block device along with file access permissions
and saves the data to external inode table file.
.SH "SYNOPSIS"
.BI "build_it [\-qvVnsafp] " "directory inode_table"

.SH "DESCRIPTION"

.B build_it 
uses 
.B Linux
OS filesystem drivers to recursively read directory entries in order to obtain 
information about all of the filesystem inodes.
The information is then saved to an external inode table. The original block
device can then be mounted along with the external inode table to recover data
from the device.

For this utility to work best, the user must have the appropriate filesystem 
driver (maybe read-only) for 
.B Linux 
OS available along with 
.BR ioctl (2)
.BR FIBMAP 
syscall support.
.B FIBMAP
call requires 
.BR root
permissions.
From 0.84.9 version utility use effective identificator
if it's binary file has SUID bit set to call this syscall.
So you can run its without root permissions.

The 
.B build_it
utility has been tested on the following filesystems:
.BR EXT2FS/EXT3FS , " XFS" , " ReiserFS" , " VFAT" , " JFS" , " ISO9660" , " ntfs-3g" .

The best results were obtained from those systems with full 
.BR FIBMAP 
support.
Currently, in
.BR NTFS
driver has
.BR FIBMAP 
support in FUSE
.BR ntfs-3g
driver begining from 0.20061212-BETA version.
At that if you need mount filesystem with
.BR blkdev
option.
Please, read
.BR losetup(8)
if you need to mount FS image (not block device).

Bear in mind, that all enough small files on
.B NTFS
keeps in packed format and it reflects in inode table
as sparse-files. Also as sparse files reflects
all crypted/compressed at filesystem level files.
Look 
.BR \-s 
option description.

For
.BR ReiserFS "(and " Reiser4 )
.B build_it 
calls ioctl UNPACK to avoid "tail-packing" and get right files map.

You can use
.B -n
option if you don't want any changes at filesystems (it maybe important
for recovering) but in this case you may get not quite right inode table.

.SH "OPTIONS"
.TP
.B \-q
Don't print any messages.
.TP
.B \-v
Print some additional information.
.TP
.B \-V
Print the version number and exit.
.TP
.B \-n
Don't unpack ReiserFS and Reiser4 files.
.TP
.B \-s
Print to stdout all sparse-files list.
It used by 
.BR anyconvertfs(8) 
utility to form rescue-list with converting from 
.BR ntfs-3g .
.TP
.B \-a
Use absolute addressing on the entire drive, not its separate partition.
Keep in mind when using this option, that then to mount output inode table
you must specify the entire disk (for example, /dev/hda), and not 
the separate partition (as /dev/hda1, /dev/hdb7).
In addition, it means setting inode table blocksize to sector size (512 bytes),
because measure units of partition offsets in partition table is sectors.
.TP
.B \-f
Process other filesystems located on the partiotions of the same device as 
the specified directory filesystem. So if /dev/hda1 mounted on /mnt/root, 
/dev/hda3 on /mnt/root/home and /dev/hdb1 on /mnt/root/usr, and you
specify /mnt/root as the
.I directory,
to inode table will also include all the contents of /mnt/home, 
but the contents of /mnt/usr was still will not be included to it.
Use the option only with -a.
.TP
.B \-p
Add to inode table also all parent directories from the specified
.I directory
to the root directory.
.TP
.I directory
Directory, usually root_directory/mount_point of filesystem, 
.B build_it
will then make an inode table of this mounted filesystem.
.TP
.I inode_table
File for saving of external inode table.

.SH "USAGE EXAMPLES"
Building an external inode table of filesystem on device mounted at home
directory of user:
.br
# build_it /home/user /tmp/inode.table

Note: will better if future inode.table file will on other device (or in
directory which doesn't located in recursively processed tree)

Note: if mount point in this example /home (not /home/user) then built
external inode table will not include files from other directory than
/home/user -- it will particular inode table of device.

Note: if there are will other mount points in recursively processed trees
(e.g. /home/user/mount/point) -- it will not processed (an inode table
can't keep information from various devices)

.SH "AUTHOR"
Nikolaj Krivchenkov aka unDEFER <undefer@gmail.com>

Translation correction by Adrian Shaw

.SH "BUG REPORTS"
Messages about any problem with using
.B anyfs-tools
package send to
undefer@gmail.com

.SH "AVAILABILITY"
You can obtain the last version of package at
http://anyfs-tools.sourceforge.net

.SH "SEE ALSO"
.BR anyfs-tools(8),
.BR ioctl(2),
.BR ioctl_list(2) ,
.BR anyfs_inode_table(5)


anyfs-tools/ChangeLog
* 19.01.08      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix the bug in build_xfs with wrong distributions of blocks on leaves 
		of BNO, CNT trees.
	- Fix the bug in build_it of the prevision release.
	- 0.84.12d Release

* 12.01.08      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Add -a, -f, -p options to build_it
	- 0.84.12c Release

* 02.08.07      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix bug in build_xfs with very long extent lists.
	- 0.84.12b Release

* 27.07.07      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix anyfs building with kernel version >=2.6.19
	- Some other building fix
	- Fix new (from v0.84.10) i/o buffer bug in anysurrect.
	- Extremally optimize anysurrect.
	- Add -s option to build_it.
	- Add support of converting from ntfs-3g by anyconvertfs.
	- 0.84.12 Release

* 16.12.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix anyfs building with kernel version >=2.6.18
	- 0.84.11 Release

* 09.12.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix anysurrect not efficient i/o buffer. So it faster more than in
		2 times.
	- Some Gentoo building fix.
	- 0.84.10 Release
	
* 25.11.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix anyfs building (replacing added 64-bit division to bit operations)
	- 0.84.9a Release

* 19.11.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix broken build_xfs
	- Fix progress bar
	- Add anyfuse module
	- build_it now gets blocksize from statfs (not FIGETBSZ)
	- 0.84.9 Release

* 19.11.06       Kirill A. Korinskiy <catap@catap.ru>
        - Split deb-package in multi package

* 18.11.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix building with xfsprogs >= 2.8.11
	- Add depmod to install target in anyfs/Makefile.in
	- 0.84.8 Release

* 18.11.06       Kirill A. Korinskiy <catap@catap.ru>
        - Add debian/ for debian-like stuff

* 22.10.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Some fix in progress bar of anysurrect to raise performance.
	- Fix building with e2fsprogs-1.39
	- 0.84.7 Release

* 28.08.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- In build_e2fs added support of files with size more than 2 Gb.
	- In anyconvertfs there was fixed searching mount point of device 
		if specified relative path and device mounted.
	- 0.84.6 Release

* 17.08.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Adding libany(3) manual page.
	- 0.84.5c Release

* 14.08.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Fix building build_xfs on x86_64 platform
	- 0.84.5b Release

* 13.08.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Now it depends on xfsprogs 2.8.10
	- configure now warns about disabled features
	- 0.84.5a Release

* 07.08.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Added build_xfs. Now you can convert FS to XFS with anyfs-tools.
	- Added anyconvertfs. It is userfriendly script for automate of
		converting.
	- Improved documentation. "USAGE EXAMPLES" section was added to
		manual pages
	- anyfs-tools.pot file and russian messages (ru.po) were added.
	- Many bugfixes in ALL utilities and kernel module.
	- 0.84.5 Release

* 01.07.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Some bugfix in text_files_descr.c
	- 0.84.4 Release

* 29.06.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- No any warnings with compiling.
	- In anyfs-driver all [kv]malloc/[kv]free changed to
		any_malloc/any_free.

* 25.06.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- added GZIP file format support
	- 0.84.3 Release

* 21.06.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bugfix in archieve_BZIP2_surrect() with trailing byte.
	- 0.84.2 Release

* 07.05.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Adding and using realloc_it() in build_it.

* 20.04.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bugfix in archieve_BZIP2_surrect(). It didn't close pipes.

* 19.04.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bugfix in write_it(). It always returned 0.
	
* 07.04.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bugfix in archieve_ZIP_surrect().

* 28.03.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- added ISO9660 CD Images file format support
	- 0.84 Release

* 26.03.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bugfix in anyfs/inode.c -- anyfs stat returned st_blocks as count of file blocks but not 512-byte blocks.

* 24.03.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bugfixes in configure -- now it use pkgconfig for looking header files of libmpeg2.

* 23.03.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bugfixes in MakeFiles -- it did skip src/anysurrect/scan_pes directory in `make clean` and `make depend`.
	- 0.83.3b Release

* 18.03.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bugfixes in MakeFiles -- there wasn't -lext2fs in LDFLAGS for all utilities.
	- Bugfix in anyfs/inode.c -- there wasn't #include <asm/uaccess.h>.

* 09.03.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- There are no more kernel headers in userspace, there are ext2fs headers only ;-)
	- 0.83.3a Release

* 08.03.06      Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- Bug in build_it on filesystems which doesn't return right f_files value in statfs structure fixed.
	- Adding anysurrect-plugins(3) manual with anysurrect API descriptions.
	- Bugfix in build_it -- using f_fsid field of statfs structure in Linux is bad idea.
	- 0.83.3 Release
	
* 05.03.06	Nikolaj (unDEFER) Krivchenkov <undefer@gmail.com>
	- added MPEG2 stream support (depend on libmpeg2)
	- added program multiplex MPEG2 stream support
	- bugfix in fd_seek()
	- added MAX_PDF_STREAM_SIZE option in config.h
	- Ext2fs/Ext3fs info (block links) support
	- Not economical memory allocating in kernel module fixed
	- now you may change list of recovering files without recompile with -t option

anyfs-tools/src/build_it.c
/*
 * build_it.c
 * Copyright (C) 2005-2006 Nikolaj Krivchenkov aka unDEFER <undefer@gmail.com>
 *
 * From filefrag from e2fsprogs, Copyright 2003 by Theodore Ts'o.
 * 
 * From geometry.c from  lilo-22.7.2, 
 * 	Copyright 1992-1998 Werner Almesberger.
 * 	Copyright 1999-2005 John Coffman.
 *
 */

#define _LARGEFILE64_SOURCE

#include "any.h"
#include "super.h"
#include "new_inode.h"

#include <sys/types.h>
#include <dirent.h>
#include <stdarg.h>
#include <sys/vfs.h>
#include <sys/stat.h>
#include <unistd.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#else
extern char *optarg;
extern int optind;
#endif
#include <sys/ioctl.h>
#include <linux/fd.h>
#include <linux/iso_fs.h>

#include "bitops.h"
#include "progress.h"
#include "version.h"

const char * program_name = "build_it";

struct progress_struct progress;

mode_t dir_umask = 0002;

int verbose = 0;
int quiet = 0;
int unpack = 1;
int print_sparse_files = 0;

int absolute_address = 0;
int other_filesystems = 0;
int full_path = 0;

long fs_type = 0;

const char *path;
const char *inode_table;

struct hd_geometry hd_geom;
dev_t hd_geom_st_dev = 0;

static void usage(void)
{
	fprintf(stderr, _("Usage: %s [-qvVnsafp] directory inode_table\n"),
			program_name);
	exit(1);
}

struct hard_link {
	ino_t		hl_ino;
	ino_t		hl_any_ino;
	nlink_t		hl_nlink_founded;
	struct hard_link *hl_next;
};

char *concat_strings(int n, ...);

void fill_anyinode_by_statinfo(struct any_inode *any_inode, 
		struct stat64 *stat_info) {
	any_inode->i_mode	 = stat_info->st_mode;
	any_inode->i_uid	 = stat_info->st_uid;
	any_inode->i_gid	 = stat_info->st_gid;
	any_inode->i_size	 = stat_info->st_size;
	any_inode->i_atime	 = stat_info->st_atime;
	any_inode->i_ctime	 = stat_info->st_ctime;
	any_inode->i_mtime	 = stat_info->st_mtime;
	any_inode->i_links_count = stat_info->st_nlink;
}

struct hd_geometry {
	unsigned char heads;
	unsigned char sectors;
	unsigned short cylinders;
	unsigned long start;
};


#define FIBMAP     _IO(0x00,1)  /* bmap access */
#define FIGETBSZ   _IO(0x00,2)  /* get the block size used for bmap */

#define HDIO_GETGEO             0x0301  /* get device geometry */

#ifndef REISERFS_SUPER_MAGIC
#define REISERFS_SUPER_MAGIC 0x52654973
#endif

#ifndef FUSE_SUPER_MAGIC
#define FUSE_SUPER_MAGIC 0x65735546
#endif

#if 0
#ifndef REISERFS_SUPER_MAGIC_STRING
#define REISERFS_SUPER_MAGIC_STRING "ReIsErFs"
#endif
#endif

#ifndef REISERFS_IOC_UNPACK
#define REISERFS_IOC_UNPACK             _IOW(0xCD,1,long)
#endif

#ifndef REISER4_SUPER_MAGIC
#define REISER4_SUPER_MAGIC  0x52345362
 /* (*(__u32 *)"R4Sb"); */
#endif
#ifndef REISER4_IOC_UNPACK
#define REISER4_IOC_UNPACK              _IOW(0xCD,1,long)
#endif

static unsigned long get_bmap(int fd, unsigned long block, int *err)
{
	int     ret;
	unsigned int b;
	*err = 0;

	b = block;
	ret = ioctl(fd, FIBMAP, &b); /* FIBMAP takes a pointer to an integer */
	if (ret < 0) {
		if (errno == EPERM) {
			fprintf(stderr, _("No permission to use FIBMAP ioctl; must have root privileges\n"));
			exit(1);
		}
		perror("FIBMAP");
		*err = -errno;
	}
	return b;
}

static int get_geometry(int fd, struct hd_geometry *hdg)
{
	int     ret;
	int 	err = 0;

	ret = ioctl(fd, HDIO_GETGEO, hdg); /* FIBMAP takes a pointer to an integer */
	if (ret < 0) {
		if (errno == EPERM) {
			fprintf(stderr, _("No permission to use HDIO_GETGEO ioctl\n"));
			exit(1);
		}
		perror("HDIO_GETGEO");
		err = -errno;
	}
	return err;
}

int BS;

int fill_anyinodeinfo(struct any_inode *inode, const char *path, 
		struct stat64 *stat_info, struct hd_geometry hd_geom) {
	
	int r = 0;
	int err;

	if (S_ISLNK(inode->i_mode)) {
		inode->i_info.symlink = malloc (1025);
		if (!inode->i_info.symlink) {
			return -ENOMEM;
		}

		err = readlink(path, inode->i_info.symlink, 1024);
		if (err<0) {
			free (inode->i_info.symlink);
			return -errno;
		}

		inode->i_info.symlink[err] = '\0';
	} else
	if (S_ISREG(inode->i_mode)) {
		long fd;
		int bs;
		unsigned long   i, block, fr_start=0, fr_length = 0, 
				numfrags = 1, numblocks;

		fd = open(path, O_RDONLY | O_LARGEFILE);
		if (fd < 0) {
			perror("open");
			return -EIO;
		}

		if ( geteuid()!=0 && getuid()==0 )
			setreuid( geteuid(), getuid() );

		if (unpack)
		{
			if (fs_type == REISERFS_SUPER_MAGIC) {
				if ( ( r = ioctl (fd, REISERFS_IOC_UNPACK, 1) ) )
				{
					fprintf (stderr, _("Cannot unpack ReiserFS file\n"));
					close (fd);
					return -abs(r);
				}
			}
			/* Forcing reiser4 to perform tail2extent converstion */
			if (fs_type == REISER4_SUPER_MAGIC) {
				if ( ( r = ioctl (fd, REISER4_IOC_UNPACK, 1) ) )
				{
					fprintf (stderr, _("Cannot unpack Reiser4 file\n"));
					close (fd);
					return -abs(r);
				}

				/* 
				 * As we may have the situation when extent will be included
				 * into transaction, and its item(s) will not be have the real block
				 * numbers assigned, we should perform fsync() in order to guarantee,
				 * that current atom is flushed and real block numbers assigned to 
				 * the extent(s) file was converted in.
				 */
				if ( (r = fdatasync(fd) ) )
				{
					fprintf (stderr, _("Cannot perform fdatasync\n"));
					close (fd);
					return -abs(r);
				}
			}
		}

		if (ioctl(fd, FIGETBSZ, &bs) < 0) { /* FIGETBSZ takes an int */
			perror("FIGETBSZ");
			close(fd);
			return -EIO;
		}

		numblocks = (stat_info->st_size + (bs-1)) / bs;

		/* Count number of fragments */
		for (i=0; i < numblocks; i++) {
			block = get_bmap(fd, i, &r);
			if (r<0) {
				close (fd);
				return r;
			}
			
			if (!i) fr_start = block;
			
			if ( (fr_start) ? block != (fr_start+fr_length)
					: block ) {
				fr_start = block;
				fr_length = 0;
				numfrags++;
			}
			fr_length++;
		}

		inode->i_info.file_frags = malloc 
			(sizeof(struct any_file_frags));
		if (!inode->i_info.file_frags) {
			close (fd);
			return -ENOMEM;
		}

		inode->i_info.file_frags->fr_nfrags = numfrags;
		inode->i_info.file_frags->fr_frags =
			malloc(numfrags*sizeof(struct any_file_fragment));
		if (!inode->i_info.file_frags->fr_frags) {
			free (inode->i_info.file_frags);
			close (fd);
			return -ENOMEM;
		}

		int ibs = BS/bs;
		int mbs = 1;
		if (ibs<1) 
		{
			ibs = 1; 
			mbs = bs/BS;
		}

		if (ibs<1 && !absolute_address) 
		{
			fprintf(stderr, _("Blocksize returned by statfs lesser than returned FIGETBSZ\n"
					"Please, send information about this filesystem to undefer@gmail.com\n"
					"statfs blocksize: %d\n"
					"FIGETBSZ blocksize: %d\n"
					"Filesystem type: %x\n"), 
					BS, bs, (unsigned) fs_type );

			if (fs_type == FUSE_SUPER_MAGIC)
			{
				fprintf(stderr, _(
					"\n"
					"Please, use blkdev option for FUSE-filesystems\n") );
			}

			exit(1);
		};

		fr_length = 0;
		numfrags = 0;
		for (i=0; i < numblocks; i++) {
			block = get_bmap(fd, i, &r);
			if (r<0) {
				close (fd);
				return r;
			}
			if (!i) fr_start = block;
			
			if ( (fr_start) ? block != (fr_start+fr_length)
					: block ) {
				if ( fr_start%ibs )
				{
					fprintf(stderr, _("Blocksize returned by statfs greater than returned FIGETBSZ, and\n"
							"FIBMAP return start fragment not alligned to statfs blocksize\n"
							"Please, send information about this filesystem to undefer@gmail.com\n"
							"statfs blocksize: %d\n"
							"FIGETBSZ blocksize: %d\n"
							"Filesystem type: %x\n"), 
							BS, bs, (unsigned) fs_type );
					exit(1);
				}

				inode->i_info.file_frags->fr_frags[numfrags].
					fr_start = fr_start/ibs*mbs +
					( absolute_address && fr_start ? hd_geom.start : 0);
				inode->i_info.file_frags->fr_frags[numfrags].
					fr_length = (fr_length +ibs-1)/ibs*mbs;
					
				fr_start = block;
				fr_length = 0;
				numfrags++;
			}
			fr_length++;
		}
		inode->i_info.file_frags->fr_frags[numfrags].
			fr_start = fr_start/ibs*mbs +
			( absolute_address && fr_start ? hd_geom.start : 0);
		inode->i_info.file_frags->fr_frags[numfrags].
			fr_length = (fr_length +ibs-1)/ibs*mbs;

		numfrags++;
		if ( (numfrags == 1) && !fr_start && print_sparse_files )
		{
			printf("%s\n", path);
		}

		if ( geteuid()==0 && getuid()!=0 )
			setreuid( geteuid(), getuid() );

		close(fd);
		
	} else {
		unsigned long major = major(stat_info->st_rdev);
		unsigned long minor = minor(stat_info->st_rdev);

		inode->i_info.device = major<<20 | minor;
	}

	return r;
}

int count_inodes(const char *path, struct hard_link **phl, dev_t st_dev, 
		struct hd_geometry hd_geom,
		struct any_sb_info *info, uint32_t dirino, int root,
		int gress)
{
	struct hard_link *hl;
	int err;
	int r = 1;
	struct any_dir *anydir=NULL;
	struct any_dirent **panydirent=NULL;
	DIR *dir;
	struct dirent *dirent;
	char *name;
	struct stat64 stat_info;
	int count_dirs = 0;

	if (info) {
		anydir = malloc (sizeof(struct any_dir));
		if (!anydir) {
			r = -ENOMEM;
			goto out;
		}

		memset (anydir, 0, sizeof(struct any_dir));
		panydirent = &anydir->d_dirent;
	}

	dir = opendir(path);
	if (!dir) {
		r = -EIO;
		goto fail1;
	}

	err = lstat64(path, &stat_info);
	if (err<0) {
		r = -errno;
		goto fail;
	}
	
	if ( stat_info.st_dev != st_dev  ) {
		if (!other_filesystems)
			goto fail;

		int st_major = stat_info.st_dev >> 8;
		int major = st_dev >> 8;

		if (st_major != major)
			goto fail;

		if (absolute_address)
		{
			char dev[1024];

			snprintf(dev, 1024, "/tmp/anyfs_device_%d", getpid());

			mknod (dev, S_IFBLK | 0660, stat_info.st_dev);
			int fd = open (dev, O_RDONLY);
			r = get_geometry(fd, &hd_geom);
			if (r<0) {
				goto fail;
			}

			close(fd);
			unlink(dev);

			st_dev = stat_info.st_dev;
		}

	}

	while ( ( dirent = readdir(dir) ) ) {
		ino_t any_ino=0;

		progress_update(&progress, gress+r);

		if ( (strcmp(dirent->d_name, ".")==0) ||
		  (strcmp(dirent->d_name, "..")==0) ) continue;
		
		name = concat_strings(3, path, "/", dirent->d_name);
		if (!name) {
			r = -ENOMEM;
			break;
		}
		
		err = lstat64(name, &stat_info);
		if (err<0) {
			r = -errno;
			free(name);
			break;
		}

		if (S_ISDIR(stat_info.st_mode)) {
			if (info) {
				any_ino = find_first_zero_bit(
						info->si_inode_bitmap, 
						info->si_inodes);

				while ( any_ino >= info->si_inodes ) {
					realloc_it(info, info->si_inodes + 10240);

					any_ino = find_first_zero_bit(
							info->si_inode_bitmap, 
							info->si_inodes);
				}

				set_bit(any_ino, info->si_inode_bitmap);

				fill_anyinode_by_statinfo (
						info->si_inode_table+any_ino, 
						&stat_info);
			}
			
			err = count_inodes(name, phl, st_dev, hd_geom, info, any_ino, 0,
					gress+r);
			if (err<0) {
				r = err;
				free(name);
				break;
			}

			count_dirs++;
			r += err;
		}
		else {
			if ( stat_info.st_nlink>1 ) {
				struct hard_link **pnhl;
				
				int w = 0;
				if (*phl) {
					for (pnhl = phl; *pnhl; 
							pnhl = &(*pnhl)->hl_next)
				       	{
						if ( (*pnhl)->hl_ino == stat_info.st_ino)
						{
							(*pnhl)->hl_nlink_founded++;
							any_ino = (*pnhl)->hl_any_ino;
							w = 1;
							if ( (*pnhl)->hl_nlink_founded == stat_info.st_nlink ) {
								hl = (*pnhl);
								(*pnhl) = (*pnhl)->hl_next;
								free (hl);
							}
							break;
						}
					}
				}
				
				if (!w) {
					pnhl = phl;
					while (*pnhl) {
						pnhl = &(*pnhl)->hl_next;
					}
					
					hl = malloc (sizeof (struct hard_link));
					if (!hl) {
						r = -ENOMEM;
						free(name);
						break;
					}
					memset (hl, 0, sizeof (struct hard_link));

					hl->hl_ino = stat_info.st_ino;
					hl->hl_nlink_founded = 1;

					(*pnhl) = hl;

					r++;

					if (info) {
						hl->hl_any_ino = 
							any_ino = find_first_zero_bit(
									info->si_inode_bitmap, 
									info->si_inodes);

						while ( any_ino >= info->si_inodes ) {
							realloc_it(info, info->si_inodes + 10240);

							any_ino = find_first_zero_bit(
									info->si_inode_bitmap, 
									info->si_inodes);
						}

						set_bit(any_ino, info->si_inode_bitmap);

						fill_anyinode_by_statinfo (
								info->si_inode_table+any_ino, 
								&stat_info);

						err = fill_anyinodeinfo( 
								info->si_inode_table+any_ino, 
								name, &stat_info, hd_geom);

						if (err<0) {
							r = err;
							free(name);
							break;
						}
					}
				}
			}
			else {
				r++;
				
				if (info) {
					any_ino = find_first_zero_bit(
							info->si_inode_bitmap, 
							info->si_inodes);
					
					while ( any_ino >= info->si_inodes ) {
						realloc_it(info, info->si_inodes + 10240);

						any_ino = find_first_zero_bit(
								info->si_inode_bitmap, 
								info->si_inodes);
					}

					set_bit(any_ino, info->si_inode_bitmap);

					fill_anyinode_by_statinfo (
							info->si_inode_table+any_ino, 
							&stat_info);

					err = fill_anyinodeinfo( 
							info->si_inode_table+any_ino, 
							name, &stat_info, hd_geom);

					if (err<0) {
						r = err;
						free(name);
						break;
					}
				}
			}
		}

		if (info) {
			(*panydirent) = malloc (sizeof(struct any_dirent));
			if (!(*panydirent)) {
				r = -ENOMEM;
				free(name);
				break;
			}
			memset (*panydirent, 0, sizeof(struct any_dirent));

			(*panydirent)->d_name = 
				malloc (strlen(dirent->d_name)+1);
			if (!(*panydirent)->d_name) {
				r = -ENOMEM;
				free(name);
				break;
			}

			strcpy((*panydirent)->d_name, dirent->d_name);

			(*panydirent)->d_inode = any_ino;

			anydir->d_ndirents++;

			panydirent = &(*panydirent)->d_next;
		}
		
		free(name);
	}

fail:
	closedir(dir);

fail1:
	if (root)
	{
		struct hard_link *hl_next;
		hl = *phl;
		if (hl) hl_next = hl->hl_next;
		for (; hl; hl = hl_next) {
			if (info) {
				info->si_inode_table[hl->hl_any_ino].
					i_links_count =
					hl->hl_nlink_founded;
			}
			hl_next = hl->hl_next;
			free(hl);
		}
		*phl = NULL;
	}
	
	if ( info && (r<=0) ) {
		struct any_dirent *dirent;
		struct any_dirent *nextdirent;
		
		dirent = anydir->d_dirent;
		if (dirent) nextdirent = dirent->d_next;
		
		for (; dirent; dirent=nextdirent) {
			nextdirent = dirent->d_next;
			free(dirent->d_name);
			free(dirent);
		}
		
		free(anydir);
	}

	if ( info && (r>0) )
	{
		info->si_inode_table[dirino].i_info.dir = anydir;
		info->si_inode_table[dirino].i_links_count = count_dirs + 2;
	}
	
out:
	return r;
}

static void PRS(int argc, const char *argv[])
{
	int             c;
	int show_version_only = 0;

	while ((c = getopt (argc, (char**) argv,
		     "qvVnsafp")) != EOF) {
		switch (c) {
		case 'q':
			quiet = 1;
			break;
		case 'v':
			verbose++;
			break;
		case 'V':
			show_version_only++;
			break;
		case 'n':
			unpack = 0;
			break;
		case 's':
			print_sparse_files = 1;
			break;
		case 'a':
			absolute_address = 1;
			break;
		case 'f':
			other_filesystems = 1;
			break;
		case 'p':
			full_path = 1;
			break;
		default:
			usage();
		}
	}

	if ( ( optind >= (argc-1) ) && !show_version_only)
		usage();

	if (!quiet || show_version_only)
		fprintf (stderr, "build_it %s (%s)\n", ANYFSTOOLS_VERSION,
				ANYFSTOOLS_DATE);

	if (other_filesystems && !absolute_address)
	{
		fprintf (stderr, "-f option may be used only with -a option\n");
		exit(0);
	}

	if (show_version_only) {
		exit(0);
	}
	
	path = argv[optind++];
	inode_table = argv[optind++];
}

int main(int argc, const char *argv[])
{
	int r;
	int err;
	int c, ac;
	struct any_sb_info *info;
	
	struct stat64 stat_info;
	struct statfs statfs_info;
	
	struct hard_link *hl = NULL;
	dev_t st_dev;
	uint32_t rootdir = 1;

#ifdef ENABLE_NLS
	setlocale(LC_MESSAGES, "");
	setlocale(LC_CTYPE, "");
	bindtextdomain(NLS_CAT_NAME, LOCALEDIR);
	textdomain(NLS_CAT_NAME);
#endif

	PRS(argc, argv);
	
	err = lstat64(path, &stat_info);
	if (err<0) {
		r = -errno;
		perror("lstat");
		goto out;
	}

	st_dev = stat_info.st_dev;
	
	err = statfs(path, &statfs_info);
	if (err<0) {
		r = -errno;
		perror("statfs");
		goto out;
	}

	fs_type = statfs_info.f_type;

	if (!S_ISDIR(stat_info.st_mode)) {
		r = -1;
		fprintf (stderr, _("\"%s\" -- Is not directory\n"),
				path);
		goto out;
	}

	if (full_path)
	{
		err = chdir(path);
		if (err<0) {
			r = -errno;
			perror("chdir");
			goto out;
		}

		path = malloc(1024);
		path = getcwd((char*) path, 1024);
		if (!path) {
			r = -errno;
			perror("getcwd");
			goto out;
		}

		r = mkpathino((char*) path, rootdir, info, &rootdir);
		if (r) {
			fprintf (stderr, 
				_("Error while mkpathino root directory\n") );
			goto out;
		}
	}


#if 0
	if (verbose)
		printf (_("start count inodes\n"));

	if (quiet)
		memset(&progress, 0, sizeof(progress));
	else
		progress_init(&progress, _("count inodes: "), 0);

	c = count_inodes(path, &hl, st_dev, hd_geom, NULL, 1, 1, 0);
	if (c<0) {
		r = c;
		goto out;
	}

	progress_update(&progress, c);
	progress_close(&progress);

	if (verbose)
	{
		printf (_("founded %d inodes\n"), c);
		printf (_("creating inode table\n"));
	}
	
	ac = c+1024;
#else
	if (verbose)
		printf (_("creating inode table\n"));

	c = statfs_info.f_files - statfs_info.f_ffree;
	ac = statfs_info.f_files - statfs_info.f_ffree
		+ statfs_info.f_bfree/1024 + 256;
	if ( !statfs_info.f_files )
		ac = statfs_info.f_blocks/1024 + 256;

	if (other_filesystems) c = 0;
#endif

	BS = statfs_info.f_bsize;

	if (absolute_address)
		BS=512;

	r = alloc_it(&info, BS, ac);
	if (r<0) goto out;

	set_bit(1, info->si_inode_bitmap);
	fill_anyinode_by_statinfo (info->si_inode_table+1, &stat_info);
	
	if (quiet)
		memset(&progress, 0, sizeof(progress));
	else
		progress_init(&progress, _("creating inode table: "),
				c);
	
	c = count_inodes(path, &hl, st_dev, hd_geom, info, rootdir, 1, 0);
	if (c<0) {
		r = c;
		goto free_out;
	}

	if (full_path)
	{
		free((void*) path);
	}
	
	progress_close(&progress);
	
	if (verbose)
		printf (_("writing inode table\n"));
	
	r = write_it (info, (char*)inode_table);	
	if (r)
	{
		fprintf(stderr,
				_("Error while writing inode table: %s\n"),
				(errno)?strerror(errno):_("format error"));
		goto free_out;
	}

free_out:
	free_it (info);	

out:
	return r;
}
	
